sklearn/metrics/tests/test_ranking.py

```diff
@@ -391,6 +391,100 @@ def test_auc_errors():
     assert_raises(ValueError, auc, [1.0, 0.0, 0.5], [0.0, 0.0, 0.0])
 
 
+def test_multi_auc_toydata():
+    # Tests the one-vs-one multiclass ROC AUC algorithm
+    # on a small example, representative of an expected use case.
+    y_true = np.array([0, 1, 0, 2])
+    n_labels = len(np.unique(y_true))
+    y_scores = np.array(
+        [[0.1, 0.8, 0.1], [0.3, 0.4, 0.3], [0.35, 0.5, 0.15], [0, 0.2, 0.8]])
+
+    # Used to compute the expected output.
+    # Consider labels 0 and 1:
+    # positive label is 0, negative label is 1
+    score_01 = roc_auc_score([1, 0, 1], [0.1, 0.3, 0.35])
+    # positive label is 1, negative label is 0
+    score_10 = roc_auc_score([0, 1, 0], [0.8, 0.4, 0.5])
+    average_score_01 = (score_01 + score_10) / 2.
+
+    # Consider labels 0 and 2:
+    score_02 = roc_auc_score([1, 1, 0], [0.1, 0.35, 0])
+    score_20 = roc_auc_score([0, 0, 1], [0.1, 0.15, 0.8])
+    average_score_02 = (score_02 + score_20) / 2.
+
+    # Consider labels 1 and 2:
+    score_12 = roc_auc_score([1, 0], [0.4, 0.2])
+    score_21 = roc_auc_score([0, 1], [0.3, 0.8])
+    average_score_12 = (score_12 + score_21) / 2.
+
+    # Unweighted, one-vs-one multiclass ROC AUC algorithm
+    sum_avg_scores = average_score_01 + average_score_02 + average_score_12
+    ovo_unweighted_coefficient = 2. / (n_labels * (n_labels - 1))
+    ovo_unweighted_score = ovo_unweighted_coefficient * sum_avg_scores
+    assert_almost_equal(
+        roc_auc_score(y_true, y_scores, multiclass="ovo"),
+        ovo_unweighted_score)
+
+    # Weighted, one-vs-one multiclass ROC AUC algorithm
+    # Each term is weighted by the posterior for the positive label.
```

Comment:
"Posterior" seems strange here. I think "prevalence" would be an appropriate term.

But I don't really understand how this is okay. If we renamed label 0 to 3 and correspondingly reordered the score columns, we should expect the same overall score. When we take each OvO pair, the order (i.e. which label is pos and which neg) is meaningless, and should have no effect on weight. You should have a test to check this (an example to test this property below).

Rather the weighting should act as if each class has been positive and negative in each pair, so the weight should be the sum of the prevalences of pos and neg class involved in each term.

Illustrating invariance under permutation (broken atm for ovo-weighted)
```python
import numpy as np
from sklearn.metrics import roc_auc_score
y_score = np.random.rand(100, 3)
# alter class prevalence
y_score[:, 2] += .1
y_score[:, 1] -= .1
y_true = np.argmax(y_score, axis=1)
y_true[np.random.randint(len(y_score), size=20)] = np.random.randint(2, size=20)
for multiclass in ['ovr', 'ovo']:
    for average in ['macro', 'weighted']:
        print(multiclass, average)
        for perm in [[0, 1, 2], [0, 2, 1], [1, 0, 2],
                     [1, 2, 0], [2, 0, 1], [2, 1, 0]]:
            inv_perm = np.zeros(3, dtype=int)
            inv_perm[perm] = np.arange(3)
            y_score_perm = y_score[:, inv_perm]
            y_true_perm = np.take(perm, y_true)
            print(' ', roc_auc_score(y_true_perm, y_score_perm,
                                     multiclass=multiclass, average=average))
```